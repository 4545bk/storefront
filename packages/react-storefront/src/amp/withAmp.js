/**
 * @license
 * Copyright Â© 2017-2018 Moov Corporation.  All rights reserved.
 */
import { inject, observer } from 'mobx-react'
import React, { Component, Fragment } from 'react'
import { Helmet } from 'react-helmet'

/**
 * An HOC that links the AMP equivalent of the current page by adding the ink rel="amphtml" 
 * tag to the document head. Assign this to all views for which AMP is implemented
 * 
 * Example:
 * 
 *  import React, { Component } from 'react'
 *  import withAmp from 'react-storefront/amp/withAmp'
 * 
 *  @withAmp
 *  export default class Product extends Component {
 *    ...
 *  }
 * 
 * @param {React.Component} WrappedComponent The component to add
 * @return {React.Component}
 */
export default function withAmp(WrappedComponent) {

  @inject('app', 'history')
  @observer
  class WithAmp extends Component {
    render() {
      const { pathname, search, hostname } = this.props.app.location
      const { amp } = this.props.app

      return (
        <Fragment>
          {!amp && (
            <Helmet key="amp">
              <link rel="amphtml" href={`https://${hostname}${pathname}.amp${search}`}/>
            </Helmet>
          )}
          <WrappedComponent key="component" {...this.props} />
        </Fragment>
      )
    }

  }

  return WithAmp
}