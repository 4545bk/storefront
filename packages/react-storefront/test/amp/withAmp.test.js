/**
 * @license
 * Copyright Â© 2017-2018 Moov Corporation.  All rights reserved.
 */
import React from 'react'
import withAmp from "../../src/amp/withAmp"
import { mount } from 'enzyme'
import { Provider } from "mobx-react";
import { createMemoryHistory } from 'history'

describe('withAmp', () => {

  it('should add the link rel="amphtml" to the document head', () => {
    const Test = withAmp(() => <div/>)
    const history = createMemoryHistory({ initialEntries: ['/foo?bar=1']})
    const wrapper = mount(<Provider app={{ amp: false, location: { hostname: 'localhost', pathname: '/', search: '' } }} history={history}><Test/></Provider>)
    expect(wrapper).toMatchSnapshot()
  })

})