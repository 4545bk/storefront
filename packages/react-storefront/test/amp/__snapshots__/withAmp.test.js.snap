// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withAmp should add the link rel="amphtml" to the document head 1`] = `
<Provider app={{...}} history={{...}}>
  <inject-WithAmp-with-app-history>
    <WithAmp app={{...}} history={{...}}>
      <HelmetWrapper defer={true} encodeSpecialCharacters={true}>
        <SideEffect(NullComponent) defer={true} encodeSpecialCharacters={true} link={{...}}>
          <NullComponent defer={true} encodeSpecialCharacters={true} link={{...}} />
        </SideEffect(NullComponent)>
      </HelmetWrapper>
      <Component app={{...}} history={{...}}>
        <div />
      </Component>
    </WithAmp>
  </inject-WithAmp-with-app-history>
</Provider>
`;
